amaranth/vendor/_gatemate.py

```diff
@@ -0,0 +1,95 @@
+from abc import abstractmethod
+from amaranth import *
+from amaranth.build import *
+from amaranth.lib.cdc import ResetSynchronizer
+
+
+
+__all__ = ["GateMatePlatform"]
+
+
+
+class GateMatePlatform(TemplatedPlatform):
+    """
+
+    Required tools:
+        * ``yosys``
+        * ``p_r``
+
+    The environment is populated by running the script specified in the environment variable
+    ``AMARANTH_ENV_GATEMATE``, if present.    
+    """
+
+    device = property(abstractmethod(lambda: None))
+    package = property(abstractmethod(lambda: None))
+
+    toolchain = "GateMate"
+
+    required_tools = [
+        "yosys",
+        "p_r",
+    ]
+
+    file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.ys": r"""
+            # {{autogenerated}}
+            {% for file in platform.iter_files(".v") -%}
+                read -sv {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
```

Comment:
This isn't right: you should not be reading `.v` files in SystemVerilog mode. You should copy this part from e.g. SiliconBlue platform.

Reply:
After reading all user-provided files, you should read the top-level `.il` file you should write above. This can again be copied from SiliconBlue essentially as-is.

Reply:
The .ys script is now:

```
"{{name}}.ys": r"""
            # {{autogenerated}}
            {% for file in platform.iter_files(".v") -%}
                read_verilog {{get_override("read_verilog_opts")|options}} {{file}}
            {% endfor %}
            {% for file in platform.iter_files(".sv") -%}
                read_verilog -sv {{get_override("read_verilog_opts")|options}} {{file}}
            {% endfor %}
            {% for file in platform.iter_files(".il") -%}
                read_ilang {{file}}
            {% endfor %}
            read_ilang {{name}}.il
            {{get_override("script_after_read")|default("# (script_after_read placeholder)")}}
            synth_gatemate {{get_override("synth_opts")|options}} -top {{name}} -vlog {{name}}_synth.v
            {{get_override("script_after_synth")|default("# (script_after_synth placeholder)")}}
        """,
```

In the documentation provided here:
![image](https://github.com/user-attachments/assets/721629ce-4836-4bf6-989f-3b9dfe6e5aaa)
There is an option to write json, but is supported for a future release of the `p_r` tool with `nextpnr` that is planned next year.

Reply:
The `command_template` for yosys is the same for the one in SilliconBlue:

```
        r"""
        {{invoke_tool("yosys")}}
            {{quiet("-q")}}
            {{get_override("yosys_opts")|options}}
            -l {{name}}.rpt
            {{name}}.ys
        """,
```

But running the code for my Blinky example I get this:
```
python3 Blinky.py -b -p amaranth_boards.gatemate_a1_evb.GateMate_A1_EVB
ERROR: Module `top' not found!
Traceback (most recent call last):
  File "/home/user/FPGA/tools/amaranth/examples/gatemate/Blinky.py", line 125, in <module>
    plat.build(Blinky(num_leds, clock_divider), do_program=do_program)
  File "/home/user/FPGA/tools/amaranth/amaranth/build/plat.py", line 103, in build
    products = plan.execute_local(build_dir)
  File "/home/user/FPGA/tools/amaranth/amaranth/build/run.py", line 118, in execute_local
    subprocess.check_call(["sh", f"{self.script}.sh"],
  File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['sh', 'build_top.sh']' returned non-zero exit status 1.

```

Reply:
Yes, at the moment `-vlog` is the correct one.

Reply:
> But running the code for my Blinky example I get this:

That is odd. If you push your changes I can look into it.

Reply:
I pushed the changes.

Reply:
I added my `Blinky.py` example aswell in `examples` for now so you can see.
Will delete later.